server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java

```diff
@@ -315,6 +338,37 @@ public void incomingRequest(final HttpRequest httpRequest, final HttpChannel htt
         }
     }
 
+    void updateClientStats(final HttpRequest httpRequest, final HttpChannel httpChannel) {
+        HttpStats.ClientStats clientStats = httpChannelStats.get(System.identityHashCode(httpChannel));
+        if (clientStats != null) {
+            if (clientStats.agent == null) {
+                if (httpRequest.getHeaders().containsKey("x-elastic-product-origin")) {
+                    clientStats.agent = httpRequest.getHeaders().get("x-elastic-product-origin").get(0);
+                } else if (httpRequest.getHeaders().containsKey("User-Agent")) {
+                    clientStats.agent = httpRequest.getHeaders().get("User-Agent").get(0);
+                }
+            }
+            if (clientStats.localAddress == null) {
+                clientStats.localAddress = NetworkAddress.format(httpChannel.getLocalAddress());
+                clientStats.remoteAddress = NetworkAddress.format(httpChannel.getRemoteAddress());
+            }
+            if (clientStats.forwardedFor == null) {
+                if (httpRequest.getHeaders().containsKey("x-forwarded-for")) {
+                    clientStats.forwardedFor = httpRequest.getHeaders().get("x-forwarded-for").get(0);
+                }
+            }
+            if (clientStats.opaqueId == null) {
+                if (httpRequest.getHeaders().containsKey("x-opaque-id")) {
+                    clientStats.opaqueId = httpRequest.getHeaders().get("x-opaque-id").get(0);
+                }
+            }
+            clientStats.lastRequestTimeMillis = threadPool.absoluteTimeInMillis();
+            clientStats.lastUri = httpRequest.uri();
+            clientStats.requestCount++;
+            clientStats.requestSizeBytes += httpRequest.content().length();
```

Comment:
Even though these are marked as `volatile`, there is still a race condition here because we do a read-then-write that could be from multiple threads. I think instead we should use a `LongAdder` for both of these